maintainers/scripts/haskell/hydra-report.hs

```diff
@@ -0,0 +1,233 @@
+   , failedMaintainedDependencies :: Seq (BuildName, Seq Text)
+   , failedPackageBuilds :: Seq BuildName
+   , failedPackageDependencies :: Seq BuildName
+   , failedJobBuilds :: Seq BuildName
+   , failedJobDependencies :: Seq BuildName
+   }
+   deriving (Show)
+
+instance Semigroup StatusSummary where
+   (<>) lhs rhs =
+      StatusSummary
+         { successfulBuilds = successfulBuilds lhs + successfulBuilds rhs
+         , failedPackageBuilds = failedPackageBuilds lhs <> failedPackageBuilds rhs
+         , failedPackageDependencies = failedPackageDependencies lhs <> failedPackageDependencies rhs
+         , failedJobBuilds = failedJobBuilds lhs <> failedJobBuilds rhs
+         , failedJobDependencies = failedJobDependencies lhs <> failedJobDependencies rhs
+         , failedMaintainedBuilds = failedMaintainedBuilds lhs <> failedMaintainedBuilds rhs
+         , failedMaintainedDependencies = failedMaintainedDependencies lhs <> failedMaintainedDependencies rhs
+         , unfinishedBuilds = unfinishedBuilds lhs + unfinishedBuilds rhs
+         , canceledBuilds = canceledBuilds lhs + canceledBuilds rhs
+         }
+
+instance Monoid StatusSummary where
+   mempty = StatusSummary 0 0 0 mempty mempty mempty mempty mempty mempty
+
+main :: IO ()
+main = do
+   args <- getArgs
+   case args of
+      ["get-report"] -> getBuildReports
+      ["ping-maintainers"] -> printMaintainerPing
+      ["mark-broken-list"] -> printMarkBrokenList
+      _ -> putStrLn "Usage: get-report | ping-maintainers | mark-broken-list"
+
+reportFileName = "haskell-updates-build-report.json"
+
+showT :: Show a => a -> Text
+showT = Text.pack . show
+
+getBuildReports :: IO ()
+getBuildReports = runReq defaultHttpConfig $ do
+   -- GET request http response
+   now <- liftIO getCurrentTime
+   r <- req GET (https "hydra.nixos.org" /: "jobset" /: "nixpkgs" /: "haskell-updates" /: "evals") NoReqBody jsonResponse mempty
+   let eval = Seq.lookup 0 . evals $ (responseBody r :: JobsetEvals)
+   eval & maybe (liftIO $ putStrLn "No Evalution found") \eval -> do
+      (buildReports :: Seq Build) <- forM (builds eval) \buildId ->
+         responseBody <$> req GET (https "hydra.nixos.org" /: "build" /: showT buildId) NoReqBody jsonResponse mempty
+      liftIO $ encodeFile reportFileName (eval, now, buildReports)
+
+maintainersQuery = "with import ./. {}; lib.mapAttrs (_: v: map (m: m.github) v.meta.maintainers) (lib.filterAttrs (_: v: builtins.length (v.meta.maintainers or []) > 0) haskellPackages)"
+command = "nix-instantiate"
+params = ["--eval", "--strict", "--json", "-"]
+
+type MaintainerMap = Map Text (Seq Text)
+
+getMaintainerMap :: IO MaintainerMap
+getMaintainerMap = do
+   nixOutput <- readProcess command params maintainersQuery
+   pure $ fromMaybe (error "Failed to decode MaintainerMap") (decodeStrict' (ByteString.pack nixOutput))
+
+buildSummary :: MaintainerMap -> Seq Build -> StatusSummary
+buildSummary maintainerMap = foldMap toSummary
```

Comment:
This means we can't ping maintainers of top-level packages who have a maintainer? Some packages use the `haskellPackages` infrastructure, but have a dedicated nix expression outside of `haskellPackages`.

Reply:
Yeas, right now it’s that way. I still think this is a huge win. I briefly tried to get the expression to include maintainers for top-level packages as well, but I met problems there. (Also it would probably make the script still slower. But I think it would be still okay.)

Reply:
Oh, btw. This goes actually back to the question I asked you in discord. The most principled way to do this would be to write a nix-expression that takes the job list from our test job iterates over that and queries the maintainers for that. Than we can be sure that we know the maintainers for every job.

Reply:
You can easily achieve that by running `hydra-eval-jobs` and using its json output. It only returns emails for the maintainers field however, but this should be enough to find the github name in most cases.

Reply:
Aaah, great!

Reply:
Done, we now find maintainers for all jobs.